src/Microsoft.AspNetCore.Routing/Tree/TreeRouteBuilder.cs

```diff
@@ -333,7 +333,8 @@ private void AddEntryToTree(UrlMatchingTree tree, InboundRouteEntry entry)
                     continue;
                 }
 
-                if (part.IsParameter && (part.IsOptional || part.IsCatchAll || part.DefaultValue != null))
+                if (part.IsParameter && IsOptionalOrCatchAllOrHasDefaultValue(part) &&
+                    RemainingSegmentsAreOptional(entry.RouteTemplate.Segments, i))
```

Comment:
You could also just start your loop at `j` instead of `j+1` and then you don't need the check for `IsOptionalOrCatchAllOrHasDefaultValue` here